File: ObjcExample/ToggleTableViewCell.m
//
//  ToggleTableViewCell.m
//  ObjcExample
//
//  Created by Timur Ganiev on 08.12.2020.
//  Copyright © 2020 Timur Ganiev. All rights reserved.
//

#import "ToggleTableViewCell.h"

@implementation ToggleTableViewCell

#pragma mark - Public

#pragma mark Lifecycle

- (instancetype)initWithText:(NSString *)text enabled:(BOOL)enabled {
    
    self = [super initWithFrame:CGRectZero];
    
    if (self) {
        [self _configureWithText:text enabled:enabled];
    }
    
    return self;
}

#pragma mark Interface

- (void)addTarget:(id)target action:(SEL)action {
    
    UISwitch *switchView = (UISwitch *)self.accessoryView;
    
    [switchView addTarget:target action:action forControlEvents:UIControlEventValueChanged];
}

#pragma mark - Private

#pragma mark Setup

- (void)_configureWithText:(NSString *)text enabled:(BOOL)enabled {
    
    UISwitch *switchView = [[UISwitch alloc] initWithFrame:CGRectZero];
    
    [switchView setOn:enabled];
    
    self.accessoryView = switchView;
    
    if (@available(iOS 14.0, *)) {
        UIListContentConfiguration *config = [UIListContentConfiguration subtitleCellConfiguration];
        config.text = text;
        self.contentConfiguration = config;
    } else {
        self.textLabel.text = text;
    }
}

@end

File: SwiftExample/ToggleTableViewCell.swift
//
//  ToggleTableViewCell.swift
//  SwiftExample
//
//  Created by Timur Ganiev on 09.12.2020.
//  Copyright © 2020-2023 Timur Ganiev. All rights reserved.
//

import UIKit

final class ToggleTableViewCell: UITableViewCell {

    // MARK: Private Properties

    private var switchControl: UISwitch? {
        accessoryView as? UISwitch
    }

    private var toggleHandler: ((Bool) -> Void)?

    // MARK: Lifecycle

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        setup()
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { nil }
}

// MARK: Internal Methods

extension ToggleTableViewCell {
    
    func render(with item: ToggleItem, toggleHandler: @escaping (Bool) -> Void) {

        if #available(iOS 14.0, *) {
            var config = UIListContentConfiguration.subtitleCell()
            config.text = item.title
            contentConfiguration = config
        } else {
            textLabel?.text = item.title
        }

        switchControl?.setOn(item.isOn, animated: false)

        self.toggleHandler = toggleHandler
    }
}

// MARK: Private Methods

private extension ToggleTableViewCell {

    func setup() {
        let switchControl = UISwitch(frame: .zero)
        switchControl.addTarget(target, action: #selector(handleToggle), for: .valueChanged)
        accessoryView = switchControl
    }

    @objc
    func handleToggle() {
        toggleHandler?(switchControl?.isOn ?? false)
    }
}
